sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,315 @@
+        if callable(self.distance_metric):
+            self.distance_func = self.distance_metric
+        elif self.distance_metric in PAIRWISE_DISTANCE_FUNCTIONS:
+            self.distance_func = \
+                PAIRWISE_DISTANCE_FUNCTIONS[self.distance_metric]
+        else:
+            raise ValueError("distance_metric needs to be " +
+                             "callable or one of the " +
+                             "following strings: " +
+                             "%s. Instead, '%s' was given"
+                             % (PAIRWISE_DISTANCE_FUNCTIONS.keys(),
+                                self.distance_metric))
+
+        # Check clustering_method
+        if self.clustering_method not in self.CLUSTERING_METHODS:
+            raise ValueError("clustering must be one of the following: %s"
+                             % self.CLUSTERING_METHODS)
+
+        # Check init
+        if self.init not in self.INIT_METHODS:
+            raise ValueError("init needs to be one of " +
+                             "the following: " +
+                             "%s" % self.INIT_METHODS)
+
+        # Check random state
+        self.random_state_ = check_random_state(self.random_state)
+
+    def fit(self, X, y=None):
+        """Fit K-Medoids to the provided data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape=(n_samples, n_features)
+
+        Returns
+        -------
+        self
+        """
+
+        self._check_init_args()
+
+        # Check that the array is good and attempt to convert it to
+        # Numpy array if possible
+        X = self._check_array(X)
+
+        # Apply distance metric to get the distance matrix
+        D = self.distance_func(X)
+
+        medoid_ics = self._get_initial_medoid_indices(D, self.n_clusters)
+        cluster_ics = self._get_cluster_ics(D, medoid_ics)
+
+        # Old medoids will be stored here for reference
+        old_medoid_ics = np.zeros((self.n_clusters,))
+
+        # Continue the algorithm as long as
+        # the medoids keep changing and the maximum number
+        # of iterations is not exceeded
+        self.n_iter_ = 0
+        while not np.all(old_medoid_ics == medoid_ics) and \
+                self.n_iter_ < self.max_iter:
+            self.n_iter_ += 1
+
+            # Keep a copy of the old medoid assignments
+            old_medoid_ics = np.copy(medoid_ics)
+
+            # Get cluster indices
+            cluster_ics = self._get_cluster_ics(D, medoid_ics)
+
+            # Update medoids with the new cluster indices
+            self._update_medoid_ics_in_place(D, cluster_ics, medoid_ics)
+
+        # Expose labels_ which are the assignments of
+        # the training data to clusters
+        self.labels_ = cluster_ics
+
+        # Expose cluster centers, i.e. medoids
+        self.cluster_centers_ = X[medoid_ics]
+
+        # Return self to enable method chaining
+        return self
+
+    def _check_array(self, X):
```

Comment:
Do you mean ClusterMixin? I'm not sure if it's worth it, though. It would be used in K-Means and K-Medoids only (unless I'm missing something). That would mean we should add something specific for K-Means and K-Medoids.